emote/optimizers.py

```diff
@@ -0,0 +1,95 @@
+# Adapted from
+# https://github.com/karpathy/minGPT/blob/3ed14b2cec0dfdad3f4b2831f2b4a86d11aef150/mingpt/model.py#L136
+# (MIT license)
+
+from typing import List, Optional, Set, Tuple
+
+import torch
+
+
+def separate_modules_for_weight_decay(
+    network: torch.nn.Module,
+    whitelist_weight_modules: Tuple[torch.nn.Module],
+    blacklist_weight_modules: Tuple[torch.nn.Module],
+    layers_to_exclude: Optional[List[str]] = None,
+) -> Tuple[Set, Set]:
+    """Separate the parameters of network into two sets: one set of parameters that will have weight decay, and one set that will not.
+
+    Args:
+        network (torch.nn.Module): Network whose modules we want to separate.
+        whitelist_weight_modules (Tuple[torch.nn.Module]): Modules that should have weight decay applied to the weights.
+        blacklist_weight_modules (Tuple[torch.nn.Module]): Modules that should not have weight decay applied to the weights.
+        layers_to_exclude (Optional[List[str]], optional): Names of layers that should be excluded. Defaults to None.
+
+    Returns:
+        Tuple(Set, Set): Sets of modules with and without weight decay.
+    """
+
+    decay = set()
+    no_decay = set()
+
+    for mn, m in network.named_modules():
+        for pn, p in m.named_parameters():
+            fpn = "%s.%s" % (mn, pn) if mn else pn
+
+            if layers_to_exclude is not None and mn in layers_to_exclude:
+                # Weights of excluded layers will NOT be weight decayed
+                no_decay.add(fpn)
+            elif pn.endswith("bias"):
+                # Biases will not be decayed
+                no_decay.add(fpn)
+            elif pn.endswith("weight") and isinstance(m, blacklist_weight_modules):
+                # Weights of blacklist modules will NOT be weight decayed
+                no_decay.add(fpn)
+            elif pn.endswith("weight") and isinstance(m, whitelist_weight_modules):
```

Comment:
I don't understand the usage of `isinstance` here! Either the type annotation lies or the code is wrong. You're taking a list of objects, but you're checking if m is a type. Should `whitelist_weight_modules` be `tuple[Type[nn.Module]]`?